---
title: 'this is manually overridden'
date: "`r format(Sys.time(), '%d %B, %Y')`"
toc: false
fontsize: 10pt
papersize: a4 
geometry: margin=2cm
#     - \usepackage{setspace}
#     - \doublespacing
#     - \usepackage{lineno}
#     - \linenumbers
header-includes:
  - \usepackage{float}
  - \title{\fontseries{m}\selectfont Easily phylotyping \textit{E. coli} via the EzClermont web app and command-line tool.}
  - \author{\small Nicholas R. Waters,$^{1,2}$ Florence Abram,$^{1}$ Fiona Brennan,$^{1,3}$ Ashleigh Holmes,$^{4}$ and Leighton Pritchard$^{2,5\ast}$ \newline \newline \small{\textit {$^{1}$Department of Microbiology, School of Natural Sciences, National University of Ireland, Galway, Ireland}} \newline \small{\textit {$^{2,5}$Information and Computational Sciences, James Hutton Institute, Invergowrie, Dundee DD2 5DA, Scotland}} \newline \small{\textit {$^{3}$Soil and Environmental Microbiology, Environmental Research Centre, Teagasc, Johnstown Castle, Wexford, Ireland}}\newline \small{\textit {$^{4}$Cell and Molecular Sciences, James Hutton Institute, Invergowrie, Dundee DD2 5DA, Scotland}}\newline  \small{\textit {$^{5}$Strathclyde Institute of Pharmacy and Biomedical Sciences, University of Strathclyde, Glasgow, G1 1XQ, Scotland}}\newline \footnotesize{$^\ast$To whom correspondence should be addressed$:$ leighton.pritchard@strath.ac.uk \newline \newline}}
output:
  bookdown::pdf_document2: 
     fig_caption: true
     keep_tex: true
editor_options:
  chunk_output_type: console
#bibliography: /Users/nick/GitHub/soil-persistent-ecoli/soil_persistent_ecoli.bib
abstract: The Clermont PCR method for phylotyping *Escherichia coli* remains a useful classification scheme even though genome sequencing is now routine, and higher-resolution sequence typing schemes are now available. Relating preset-day whole-genome *E. coli* classifications to legacy phylotyping is essential for harmonising the historical literature and understanding of this important organism. We therefore present EzClermont - a novel *in silico* Clermont PCR phylotyping tool to enable ready application of this phylotyping scheme to whole genome assemblies. We evaluate this tool against phylogenomic classifications, and an alternative software implementation of Clermont typing. EzClermont is available as a web app at  <http://www.ezclermont.org>, and as a command-line tool at <https://nickp60.github.io/EzClermont/>.  

---
```{r, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning=FALSE, message=FALSE)
#knitr::opts_chunk$set(fig.pos = 'H')
library(tidyverse)
library(kableExtra)
```
## Introduction {-}
*Escherichia coli* is one of the the most widely-studied and best understood organisms in biology. Even before widespread whole-genome sequencing it was known that the *E. coli* species group is very diverse [@lukjancenko_comparison_2010; @selander_genetic_1987], and several methods were developed to differentiate the various  *E. coli* lineages.
<!-- %In addition to being a member of the human gut microbiome, *E. coli* are found in soil, water, and livestock. The majority of lineages are harmless, but some can cause serious illness.  Thus, detecting and identifying the different lineages has remainedva valuable goal both for understanding the pathalogy as well as for detecting problematic strains. -->
In 1987, Selandar  and colleagues first used electrophoretic analysis of a 35 enzyme digest to classify the *E. coli* Reference Collection (ECOR) into 6 groups (A-F) [@selander_genetic_1987].  Subsequently, Clermont & colleagues published a triplex PCR method for phylotyping, able to differentiate four of these groups - A, B1, B2 and D [-@clermont_rapid_2000].  In 2013, Clermont and colleagues updated this scheme, adding a fourth set of primers to detect groups E and F; additional primers were also proposed to differentiate the cryptic clades [@clermont_clermont_2013]. This method was again recently extended to include primers that differentiate the newly-identified G phylogroup [@clermont_characterization_2019]. The Claremont quadruplex primers have been widely adopted for laboratory-based classification as the method is reliable, easy to interpret, and correctly classifies about 95\% of *E. coli* strains.

Other typing schemes developed to classify *E. coli* strains include: the Achtman 7 gene Multi Locus Sequence Typing (MLST)[@achtman_multilocus_2012;@alikhan_genomic_2018]; Michigan EcMLST [@qi_ecmlst_2004]; whole-genome MLST (<http://www.applied-maths.com/applications/wgmlst>); core-genome MLST [@debeen_core_2015]; two-locus MLST[@weissman_highresolution_2012]; and ribosomal MLST [@jolley_ribosomal_2012]. All of these sequencing-based methods classify *E. coli* with greater accuracy and to higher resolution than Clermont phylotyping. Any practical choice of approach involves trade-offs of cost and complexity against the precision offered by the methodology. The @clermont_clermont_2013 phylotyping scheme remains a popular tool for *E. coli* classification as it can be performed rapidly and inexpensively in a laboratory. In addition, this classification scheme remains useful to make comparisons of newly-sequenced isolates against historical literature, which contains many references to strains classified only by the Clermont scheme.

EzClermont was developed to bridge the gap between the traditional quadruplex primer approach to phylotyping and whole-genome seqenuce data. It provides a simple *in silico* analogue of the Clermont phylotyping approach, applied to genome assemblies. We  implemented EzClermont as both a web application for public use, and as a command-line program for local installation. A similar tool called ClermonTyping was recently published, with similar goals and functionality [@beghain_clermontyping_2018].  Here, we describe our implementation of the Clermont classification scheme in EzClermont, assess its ability to correctly assign Clermont type, relative to *E. coli* whole-phylogeny, and compare its performance to the ClermonTyping program.

## Methods {-}
### *In silico* PCR {-}

To emulate PCR *in silico*, EzClermont uses regular expressions (regexes) that represent the Clermont primer sequences to locate their potential binding sites on a sequenced genome. The sequence regions lying between these sites are taken to be the predicted amplicons, and can be evaluated for sequence composition, or presence/absence to determine Clermont phylotype.

In practice, Clermont primer sequences do not always exactly match their productive binding sites, so primer-binding sequence variability must be captured in the corresponding regexes. To represent this variability, we selected 1395 *E. coli* genomes from EnteroBase[**REF**] (accessed April, 2019), filtering on genome metadata quality and source [**MORE (BUT BRIEF) DETAIL/PRECISION NEEDED - PUT DETAILED METHOD IN SUPPLEMENTARy IF NECESSARY**]. The theoretical amplicons of each of the quadruplex, E-specific, C-specific, G-specific, and E/C control primer sets were identified and aligned [**PUT DETAIL - RECIP. BLAST, MAFFT ETC. IN SUPPLEMENTARY**] to identify variations in the canonical primer regions. These were used to establish the regular expressions used to search for primer binding sites in whole genomes [**INSERT A TABLE WITH THESE REGEX SEQUENCES**]. Sequence differences in the last five 5` bases of the reverse primer in each set were not included, as these variations can be used to differentiate alleles [@stadhouders_effect_2010] [**I AM NOT YET CLEAR WHAT THEY WERE NOT INCLUDED/INCORPORATED IN - DO THE PRIMER REGEXES STOP SHORT, BY 5bp?**].

### Validation Dataset {-}

Determining the efficacy of our approach required a set of isolates with both experimentally-determined Clermont phylotypes and available whole-genome sequencing data. A dataset was created from the strains described in @sims_wholegenome_2011, @clermont_guide_2015, and some of group G isolates from @clermont_characterization_2019 for which data was available. The accessions for these 125 strains can be found in Supplementary Materials^[See file `docs/analysis/validate/validation_metadata.csv`. Note that an earlier version of this work used ECOR sequences from BioProject PRJNA321606; these sequences were found to be of poor quality, a finding supported by @patel_draft_2018. This work reports the analysis of ECOR sequences from PRJNA230969.]. 

The phylogeny of these strains were estimated by generating a core-genome alignment with Parsnp [@treangen_harvest_2014] in a manner comparable to that described by @clermont_characterization_2019.  PhyML [@guindon_new_2010] was used on the resulting nucleotide alignment using a HKY85 substitution model, bootstrapped with approximate Bayes support.  The ggtree [@yu_ggtree_2017] package was used to visualize the tree; the branches were rotated such that the branches representing the cryptic *Escherichia* initiate the tree.


### Comparison EzClermont to ClermonTyping {-}

Both EzClermont (version 0.6.2) and ClermonTyping (version 1.4.1) were run with default parameters on the 125 strains in the validation dataset to assess the performace of both tools.

### Assessing results {-}
For instances where both methods of *in silico* phylotyping predicted a different phylogroup from the reported one, the short read data was downloaded from NCBI.  Blobtools [@laetsch_blobtools_2017] was run via ezBlobtools^[<https://hub.docker.com/r/nickp60/ezblobtools/>, digest `f64d38ffce`],  on the forward read of the dataset both to visualize potential contamination; the resulting assembly performed with SKESA [@souvorov_skesa_2018] was also re-assessed with the in-silico phylotypers.

### Comparing the ECOR sequencing projects {-}
The assemblies from PRJNA321606 and  PRJNA230969 were downloaded; a core-genome alignment and SNP tree were made with Parsnp using default parameters.  The tree was visualized and assessed based on whether a strain's assemblies ended up clustering along with each other.  

### Implementation {-}
 
EzClermont is a Python package available on PyPI, conda, and GitHub.  The package consists of a command-line tool for batch execution, and a Flask-based webapp.  This webapp is also hosted on <http://ezclermont.org>, and a Docker container is available for those wishing to deploy it locally.


## Results {-}

Before addressing the performance of the *in silico* phylotypers, the experimentally-determined phylogroups must be confirmed with the maximum likelihood phylogeny.  Of the 125 strains assessed, seven of the reported phylogroup do not match.  Table 1 summarizes any disagreement between the literature reported phylotypes and that determined by maximum likelihood, ClermonTyping, or EzClermont. This could be for a number of reasons, including a strain being mislabelled, contamination, or a case where the Clermont quadruplex PCR does not accurately reflect the phylogeny.  In all cases, both *in silico* methods typed the strain in a manner matching the phylogeny, supporting the maximum likelihood phylogeny and suggesting that the integrity of the strain collection may have been compromised.  Indeed, comparing the two sequencing efforts revealed that at least in these two collections, the phylogentic identities of 12 of the 72 strains are not certain (Supplementary Figure \@ref(fig:bioproject-comparison)), and research labs may be refering to different strains by the same name. For assessing performance of the *in silico* tools, the phylogroup supported by the phylogeny was taken to be true, as we have no way of knowing the true identity of the strain on which the original Clermont PCR method was performed.

```{r tree-ez, fig.cap="Cladogram of phylogenetic relationships between members of the ECOR collection and select phylogroup G isolates from Clermont et al. 2019. Clades are coloured by phylogroup; the heatmap surrounding the tree shows the reported phylogroup, ClermonTyping phylogroup, and EzClermont phylogroup (inside to outside).  The reported phylogroup was not supported by the phylogeny in seven of the strains. Both EzClermont and ClermonTyping show agreement with the phylogeny in all but two cases: ECOR44 and ECOR49.", out.width="110%" }
knitr::include_graphics(path =  "../analysis/cladogram.png")
```


The cladogram in Figure \@ref(fig:tree-ez) highlights two isolates (ECOR44 and ECOR49) were mistyped by both EzClermont and ClermonTyping; the maximum likelihood tree is shown in Supplementary Figure \@ref(fig:tree-ez-phy). Both of these mistyped  isolates  are phylogroup D.  Both tools type ECOR49 as phylogroup G. It appears that the canonical arpA fragment that should be present in phylogroup D isolates could not be identified in the assembly with either tool. Using a reciprocal BLAST approach to identify the region in the assembly also failed. Taken together, this suggests that the assembly may be incomplete, as the fragment must be present in the organism to be detected with PCR, but cannot be found; this was confirmed by running on the alternative assembly accession GCA_002190975.1, which both tools type as phylogroup D.

ClermonTyping types ECOR44  as phylogroup E, suggesting a false positive with the Group E primer set. EzClermont types it as phylogroup G, suggesting a false negative of the arpA primer set. Investigation revealed that the quadruplex's arpA fragment was not detected due to a G->A mutation in base 17 of the reverse primer site. Because this occurs in the final 5 bases of the 3' end, it was not incorporated during the training process; a similar mutation was seen in 9 of the 1395 training isolates. @beghain_clermontyping_2018 notes the difficultly in typing with this arpA fragment, as it has likely been horizontally transfered to some phylogroup D isolates.

<!-- Table (\#tab:sims)1: Summary of phylogroup by method. EzClermont and ClermonTyping were run on a set of strains with reported phylotypes. PhyML was used to reconstruct the phylogeny based on core SNPs from Parsnp, allowing comparison between the phylotype and the true phylogeny. Both tools type ECOR49 types as phylogroup G due to a contaminated assembly (*); ECOR49 from assembly GCA_002190975.1 is correctly typed by both tool s as phylogroup D. -->

```{r simstab}
restab <- structure(list(Strain = c("SMS-3-5", "APEC01", "ECOR07", "ECOR72", 
"ECOR71", "ECOR43", "ECOR44", "ECOR49", "ECOR23"), Accession = c("GCA_000019645.1", 
"GCA_003028815.1", "GCA_003334305.1", "GCA_003334425.1", "GCA_003333385.1", 
"GCA_003333775.1", "GCA_003333765.1", "GCA_003333685.1", "GCA_003334095.1"
), Reported = c("D", "B2", "A", "B1", "B1", "A", "D", "D", "A"
), Phylogeny = c("F", "A", "B1", "B1", "C", "E", "D", "D*", "B2"
), ClermonTyping = c("F", "A", "B1", "C", "C", "E", "E", "G*", 
"B2"), EzClermont = c("F", "A", "B1", "C", "C", "E", "G", "G*", 
"B2"), Note = c("", 
"", "", "", "", "", "ArpA1_r G17A", 
"", "")), row.names = c(NA, -9L), class = "data.frame")
knitr::kable( longtable=FALSE, booktabs=TRUE,
  #format = "markdown",
  escape = T, col.names = c( "Strain", "Accession",  "Reported",  "Phylogeny", " ClermonTyping", "EzClermont", "Note")  , 
  caption="Summary of phylogroup by method. EzClermont and ClermonTyping were run on a set of strains with reported phylotypes. PhyML was used to reconstruct the phylogeny based on core SNPs from Parsnp, allowing comparison between the phylotype and the true phylogeny. Both tools type ECOR49 types as phylogroup G due to a contaminated assembly (*); ECOR49 from assembly GCA002190975.1 is correctly typed by both tool s as phylogroup D.",
  restab %>% arrange(Strain)
) %>% kable_styling(latex_options = "striped")
```


## Discussion {-}

EzClermont was built to bridge the gap between traditional methods of classifying *E. coli* and the interpretation of whole-genome sequencing data. Considering the phylogeny of the strains tested, both EzClermont and ClermonTyping correctly classify 124/125 isolates.  Further, a wide application of EzClermont by @zhou_user_2019 to representative *E. coli* strains in EnteroBase was largely in agreement with both higher-resolution sequence typing and with ClermonTyping. 

ClermonTyping has the advantage of classifying which cryptic lineage a strain belongs to, where EzClermont groups them all together. Due to the many methods that provide more accurate classification of these clades, we do not attempt to improve our method to classifiy those species. The execution time for the 125 strains tools 3 minutes 56 seconds for EzClermont; ClermonTyping executed 3 minutes 16 seconds; both tools execute in similar times via their webapp implementations. 

As a python package, EzClermont is easy for developers to integrate into existing pipelines. Given the ease of use of the web app for simple queries, its incorporation into EnteroBase [@zhou_user_2019], and the standalone speed of execution for larger batches, we hope that EzClermont will be of continued use to the scientific community.





## Supplementary Figures {-}
\renewcommand\thefigure{S\arabic{figure}}    
\setcounter{figure}{0}  


```{r bioproject-comparison, fig.cap="A core-genome alignment of the assemblies from both ECOR collection sequencing projects was created with Parsnp.  The tip shapes indicate the sequencing project. (A) shows the core genome phylogeny, and (B) shows the same but with equal branch lengths to easier visualize closely-related strains. Yellow tips indicate strains for which the two BioProject's assemblies contradict.", out.width="110%" }
knitr::include_graphics(path =  "../analysis/bioproject_comparison.png")
```


```{r tree-ez-phy, fig.cap="Maximum likelihood tree of core genome alignment showing the relationship of the isolates to their literature-reported and predicted phylotypes.  The cryptic clade isolates were omitted for clarity, and the tree was rerooted at the point where they diverged.", out.width="110%" }
knitr::include_graphics(path =  "../analysis/tree.png")
```


## References {-}





